### PR TITLE
[Tech debt] Refactor authService and remove LogInHelper

### PR DESCRIPTION
**Pull request by eliykat**
## Type of change

- [ ] Bug fix
- [ ] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective

`AuthService` is a bit of a mess because the private `logInHelper` method contains code used by every single type of authentication (e.g. MP, SSO, API, 2FA). This makes it extremely difficult to reason about what type of authentication calls what code, where new code should go, and when code can be removed.

This refactor aims to improve this by:
* separating out code that is used by all login methods and code that is only required by particular login methods
* simplifying public and private interfaces (in particular, reducing the number of nullable/optional parameters)
* moving code into other services where appropriate

### Peripheral changes
Here are the smaller/peripheral changes you should understand first.

Models:
* make `tokenRequest` an abstract base class
  * add subclasses for the different types of requests: `passwordTokenRequest`, `apiTokenRequest`, `ssoTokenRequest`. Extract specific code into the subclasses.
  * group two factor data into an object with a defined interface: `TokenRequestTwoFactor` (defined in the base `tokenRequest` file)
* `authResult`: add getters. Remove the `twoFactor` property because we can calculate this information rather than storing it.

Other services:
* `KeyConnectorService`: this now has the code that enrols a new SSO user in Key Connector, which used to be handled by `authService`. This could likely be refactored/combined with `migrateUser` (which handles converting existing users), but that's out of scope for now.
* `TwoFactorService`: a new service which handles the two factor-related operations that were previously in `AuthService`. This reduces the amount that `AuthService` is responsible for given that it's already quite complex.
* `TokenService`: this interface did not match the implementation. I assume this is an oversight from account switching. I've updated the interface accordingly.
* `ApiService`: update to use the new `tokenService` interface. Remove the null coalescing when creating the identity token string, this is handled by the subclasses.

CLI
* `login.command.ts`: this used to be a 6-pronged conditional, which called either the standard login methods (`LogIn`, `LogInApi`, `LogInSso`) if no 2FA was provided, or the `*Complete` version of those methods if 2FA was provided. This has been simplified to just call those 3 standard login methods, and they will handle whether or not 2FA is provided.
  * this is begging to be broken up into subclasses, but that's out of scope this time
* Note: CLI is broken at the moment due to account switching, so I still need to do some basic testing.

Angular components:
* update to use new interfaces and services

### AuthService changes

* as mentioned above:
  * strip out Key Connector logic into that service
  * strip out 2FA methods into that new service
  * remove `LogInComplete`, `LogInSsoComplete`, and `LogInApiComplete`. These were called by the CLI client if the user provided 2FA info upfront. Now we use `buildTwoFactor` to handle those situations, and we don't need separate entry methods.
* extract the logic for performing a login into 3 new classes: `apiLogInDelegate`, `passwordLogInDelegate` and `ssoLogInDelegate`, plus their common base class. The base class has the common logic and the subclasses provide their own implementations as required, depending on the login method used. This is where the `LogInHelper` code has been moved to.
* `authService` is now only responsible for creating and initializing the correct delegate class and asking it to start the login process. It's essentially the factory for the different implementations.

I don't know if "delegate" is the correct verbiage here, but it's neither a service (which are long-lived and initialized during bootstrapping) nor a model (which only manipulate data and shouldn't have service dependencies). I've stuck them in the `misc` folder for this reason.

### Tests
Add tests. My first tests were very broad and contained lots of overlapping assertions, just as guardrails while I was refactoring. Once I finished refactoring, I rewrote them to test the specific behaviour that is expected of each delegate.

## Other clients
* cli: https://github.com/bitwarden/cli/pull/448
* desktop: https://github.com/bitwarden/desktop/pull/1239
* web: https://github.com/bitwarden/web/pull/1387
* browser: TODO once we can update jslib
* directory-connector: https://github.com/bitwarden/directory-connector/pull/213 (note this has some real changes, not just dependency updates)

## Testing requirements

This will require really comprehensive regression testing of all login methods and combinations. Here are the code paths I can think of that I've touched:
* Logging in with a Master Password
* Logging in with an Api Key (CLI only)
* Logging in with an Organization Api Key (directory connector, gui and cli)
* Logging in with SSO:
  * existing user
  * new user provisioning flow (i.e. user does not have a Bitwarden account yet)
  * existing user with Key Connector
  * new user provisioning flow with Key Connector
* Different 2FA combinations:
  * no 2FA
  * Providing 2FA at the same time as main credentials (CLI only, when you pass it as an argument)
  * Providing 2FA as part of the normal two-step prompt
  * using a saved 2FA token ("remember me")
* CAPTCHA verification

## Before you submit

- [x] I have checked for **linting** errors (`npm run lint`) (required)
- [x] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

**Comment by eliykat** [comment]
I did some further work on this today and moved it to a more object-oriented solution (using subclassing instead of passing callbacks). The PR description above has been updated.

**Review by justindbaur — COMMENTED** [review]
Just one small thing, I don't know how aggressively we are trying to avoid `any`.

File: common/src/abstractions/twoFactor.service.ts
Comment (eliykat) [reply common/src/abstractions/twoFactor.service.ts]: I rarely like `any`, I've added an interface for this.

**Review by justindbaur — APPROVED** [review]

File: common/src/abstractions/twoFactor.service.ts
Comment (justindbaur) [reply common/src/abstractions/twoFactor.service.ts]: Same here. I know it can almost be a necessity sometimes but I always like an interface. Thanks!

**Comment by eliykat** [comment]
Made some minor changes after some basic regression testing in other clients.

File: common/src/services/twoFactor.service.ts
Comment (MGibson1) [inline common/src/services/twoFactor.service.ts:93]: Again, I think this belongs in the constructor

File: node/src/cli/commands/login.command.ts
Comment (MGibson1) [inline node/src/cli/commands/login.command.ts:20]: This class is still quite horrible, but I don't think that was the focus of your refactoring

Comment (MGibson1) [reply node/src/cli/commands/login.command.ts:20]: Please don't forget to validate DirectoryConnectors authentication when testing this

**Review by MGibson1 — CHANGES_REQUESTED** [review]
I have a few small request. Really nice job with this one @eliykat! It's SO much better

File: common/src/misc/logInDelegate/apiLogin.delegate.ts
Comment (eliykat) [reply common/src/misc/logInDelegate/apiLogin.delegate.ts]: A constructor can't be `async`, but what I've done instead is create an async static factory method. This creates the delegate object, initializes it, and returns it ready to be used. The constructor of each delegate is now `private` to prevent it from being created any other way.

File: common/src/misc/logInDelegate/logIn.delegate.ts
Comment (eliykat) [reply common/src/misc/logInDelegate/logIn.delegate.ts]: Good call

File: common/src/misc/logInDelegate/passwordLogin.delegate.ts
Comment (eliykat) [reply common/src/misc/logInDelegate/passwordLogin.delegate.ts]: Done using static factory method, see other comment for details

File: common/src/misc/logInDelegate/ssoLogin.delegate.ts
Comment (eliykat) [reply common/src/misc/logInDelegate/ssoLogin.delegate.ts]: Done using static factory method, see other comment for details

File: common/src/services/twoFactor.service.ts
Comment (eliykat) [reply common/src/services/twoFactor.service.ts:93]: This was just a copy/paste to the new service. The `init` method is used because it depends on the `i18nService` being init'd first, which is handled by the client ([example here](https://github.com/bitwarden/web/blob/master/src/app/services/services.module.ts#L84)). I don't really like how we handle this, but I'm not sure of a better pattern.

File: common/src/misc/logInDelegate/passwordLogin.delegate.ts
Comment (eliykat) [reply common/src/misc/logInDelegate/passwordLogin.delegate.ts:113]: We can, but we'd have to store the plaintext Master Password in a class property so we can hash it later in `onSuccessfulLogin`. That seems too dangerous to be worthwhile. Especially as the delegate object can persist if 2FA is required.

File: node/src/cli/commands/login.command.ts
Comment (eliykat) [reply node/src/cli/commands/login.command.ts:20]: LoginCommand is next on the hitlist, don't worry ;) especially as I have a really good understanding now of the auth flows.

But yes, I ruled it out of scope for now.

File: common/src/misc/logInDelegate/logIn.delegate.ts
Comment (eliykat) [reply common/src/misc/logInDelegate/logIn.delegate.ts:74]: It seems worse to pass in the whole object, then the method call could be doing anything with it (mutable data). If it only needs 1 property, why give it the whole object? It also more tightly couples it to `authService`.

I'm okay with just passing in the property and handling null within the `twoFactorService` though. I'm always a bit undecided whether to "only call it if you need it" or "always call it and handle the null case within the called method".

File: common/src/misc/logInDelegate/logIn.delegate.ts
Comment (eliykat) [reply common/src/misc/logInDelegate/logIn.delegate.ts:80]: You're right. Turns out that `apiService.postIdentityToken` already checks what properties are present on the raw http response and instantiates the correct object accordingly. So we don't need to check `requiresCaptcha` etc, we can just branch based on the response type. I've split this method up into `processTwoFactorResponse`, `processCaptchaResponse`, and `processIdentityTokenResponse`.

File: common/src/misc/logInDelegate/apiLogin.delegate.ts
Comment (MGibson1) [reply common/src/misc/logInDelegate/apiLogin.delegate.ts]: yeah I missed that it was due to `async`, but I still think that's the proper way to solve this

File: common/src/misc/logInDelegate/passwordLogin.delegate.ts
Comment (MGibson1) [reply common/src/misc/logInDelegate/passwordLogin.delegate.ts:113]: yeah your way is better. Please comment to that effect since (at least in my reckoning) that's not an obvious downside of moving it.

File: common/src/services/twoFactor.service.ts
Comment (MGibson1) [reply common/src/services/twoFactor.service.ts:93]: hmm I see. I suppose the right way to do that would be to try and move the `init` method in the `i18nService` as well, but I'll leave you off the hook on that one...

File: common/src/misc/logInDelegate/logIn.delegate.ts
Comment (eliykat) [reply common/src/misc/logInDelegate/logIn.delegate.ts]: We do in the bwdc cli. But I just noticed that has its own authService subclass, so I'll see if I can refactor this to remove this parameter.

File: common/src/misc/logInDelegate/logIn.delegate.ts
Comment (eliykat) [reply common/src/misc/logInDelegate/logIn.delegate.ts]: Both `setEncKey` and `setEncPrivateKey` are called from other services where there is no `AuthResult`, so I think they need to stay generalized/loosely coupled.

Also, I wouldn't want `setEncPrivateKey` to always assume that a new keypair needs to be created if `privateKey` is null, there may be other situations where that's an error.

File: common/src/misc/logInDelegate/logIn.delegate.ts
Comment (eliykat) [reply common/src/misc/logInDelegate/logIn.delegate.ts]: Done, the whole response object is now passed to `twoFactorService`

File: common/src/misc/logInDelegate/logIn.delegate.ts
Comment (eliykat) [reply common/src/misc/logInDelegate/logIn.delegate.ts]: Done, the whole response object is now passed to `tokenService`

File: common/src/misc/logInDelegate/ssoLogin.delegate.ts
Comment (eliykat) [reply common/src/misc/logInDelegate/ssoLogin.delegate.ts]: Done, agree this is a good example of reducing the long parameter list.

File: common/src/misc/logInDelegate/logIn.delegate.ts
Comment (eliykat) [reply common/src/misc/logInDelegate/logIn.delegate.ts]: Update: directory-connector now has its own `OrganizationApiLoginDelegate` subclass which overrides setting the crypto keys in the parent's methods. The `setCryptoKeys` parameter has been removed.

**Comment by eliykat** [comment]
@MGibson1 I should have addressed all your feedback, plus made some changes to support directory connector and remove the `setCryptoKeys` parameter. Might be worth checking out the directory-connector PR for context even though it's in draft: https://github.com/bitwarden/directory-connector/pull/213.

Will check I haven't broken anything for the other clients tomorrow, but you can review again in the meantime.

**Comment by eliykat** [comment]
I've done a brief regression test in all clients (except browser, which is blocked by stateService changes) and it's all looking good.

**Review by MGibson1 — DISMISSED** [review]

File: common/src/models/request/identityToken/apiTokenRequest.ts
Comment (eliykat) [reply common/src/models/request/identityToken/apiTokenRequest.ts]: 😂 That's fair enough. The real problem here - which made the comment necessary - is that we were taking an argument that we were then ignoring, just to fit the same method signature as the other `tokenRequest` classes. I've now fixed this in [this commit](https://github.com/bitwarden/jslib/pull/588/commits/8e8cb3241badbbf7bc9b60af06dc9121e0aa1ade) so that we just have a different method signature in this subclass.

**Comment by eliykat** [comment]
@MGibson1 I think I've now refactored this to within an inch of its life. Summary of further changes:
1. rename `delegate` to `strategy`
2. remove `init` and static factory method in each Strategy. They now take the credentials directly into their `logIn` method
3. get rid of all the primitive arguments and pass the credentials as objects. This lets us have a single `logIn` method in`authService`, which can switch between strategies depending on the type of credentials received.
4. resolved your comment above

I'm happier with how this is looking now.

**Review by MGibson1 — APPROVED** [review]
I'm going to tag this with a hold until we can confirm with QA on merging

**Comment by eliykat** [comment]
`rc` has been cut so I'm merging this.

**Pull request merged**
